File: eisenhower.py
# eisenhower.py
import sys, os, json, argparse, uuid
from datetime import datetime
try:
    from colorama import init, Fore, Style
    init()
    COLORS = True
except ImportError:
    COLORS = False
    Fore = Style = type('', (), {'RESET_ALL':'', 'RED':'', 'GREEN':'', 'YELLOW':'', 'CYAN':'', 'WHITE':'', 'BLACK':''})()

DATA_FILE = "tasks.json"

class Task:
    def __init__(self, title, description="", quadrant=1, task_id=None):
        self.id = task_id or str(uuid.uuid4())[:8]
        self.title = title
        self.description = description
        self.quadrant = quadrant  # 1-4
        self.created = datetime.now().isoformat()
        self.completed = False

    def to_dict(self):
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "quadrant": self.quadrant,
            "created": self.created,
            "completed": self.completed
        }

    @classmethod
    def from_dict(cls, data):
        task = cls(data["title"], data.get("description", ""), data["quadrant"], data["id"])
        task.created = data.get("created", datetime.now().isoformat())
        task.completed = data.get("completed", False)
        return task

class EisenhowerApp:
    def __init__(self):
        self.tasks = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.tasks = [Task.from_dict(t) for t in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([t.to_dict() for t in self.tasks], f, indent=2)

    def add_task(self, title, desc="", quadrant=1):
        task = Task(title, desc, quadrant)
        self.tasks.append(task)
        self.save()
        print(f"Task added: {task.title} (ID: {task.id})")

    def delete_task(self, task_id):
        for i, t in enumerate(self.tasks):
            if t.id == task_id:
                del self.tasks[i]
                self.save()
                print(f"Task {task_id} deleted.")
                return
        print(f"Task {task_id} not found.")

    def move_task(self, task_id, new_quadrant):
        for t in self.tasks:
            if t.id == task_id:
                t.quadrant = new_quadrant
                self.save()
                print(f"Task {task_id} moved to quadrant {new_quadrant}.")
                return
        print(f"Task {task_id} not found.")

    def list_tasks(self, quadrant=None):
        filtered = [t for t in self.tasks if quadrant is None or t.quadrant == quadrant]
        if not filtered:
            print("No tasks found.")
            return
        quadrants = {
            1: ("Urgent & Important", Fore.RED),
            2: ("Not Urgent but Important", Fore.GREEN),
            3: ("Urgent but Not Important", Fore.YELLOW),
            4: ("Not Urgent & Not Important", Fore.WHITE)
        }
        print("\nEisenhower Matrix")
        print("="*60)
        for q in range(1,5):
            q_tasks = [t for t in filtered if t.quadrant == q]
            label, color = quadrants[q]
            if COLORS:
                print(f"{color}{label}{Style.RESET_ALL}")
            else:
                print(label)
            if q_tasks:
                for t in q_tasks:
                    print(f"  [{t.id}] {t.title} - {t.description or 'No description'}")
            else:
                print("  (empty)")
            if q < 4:
                print("-"*40)
        print("="*60)

    def stats(self):
        counts = {1:0,2:0,3:0,4:0}
        for t in self.tasks:
            counts[t.quadrant] += 1
        print("Eisenhower Statistics")
        for q, label in [(1,"Urgent & Important"), (2,"Not Urgent but Important"),
                         (3,"Urgent but Not Important"), (4,"Not Urgent & Not Important")]:
            print(f"{label}: {counts[q]} tasks")

def main():
    app = EisenhowerApp()
    parser = argparse.ArgumentParser(description="Eisenhower Matrix Todo List")
    subparsers = parser.add_subparsers(dest="command", required=True)

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("-t", "--title", required=True)
    add_parser.add_argument("-d", "--desc", default="")
    add_parser.add_argument("-q", "--quadrant", type=int, choices=[1,2,3,4], default=1)

    list_parser = subparsers.add_parser("list")
    list_parser.add_argument("-q", "--quadrant", type=int, choices=[1,2,3,4])

    move_parser = subparsers.add_parser("move")
    move_parser.add_argument("--id", required=True)
    move_parser.add_argument("-q", "--quadrant", type=int, choices=[1,2,3,4], required=True)

    delete_parser = subparsers.add_parser("delete")
    delete_parser.add_argument("--id", required=True)

    stats_parser = subparsers.add_parser("stats")

    args = parser.parse_args()
    if args.command == "add":
        app.add_task(args.title, args.desc, args.quadrant)
    elif args.command == "list":
        app.list_tasks(args.quadrant)
    elif args.command == "move":
        app.move_task(args.id, args.quadrant)
    elif args.command == "delete":
        app.delete_task(args.id)
    elif args.command == "stats":
        app.stats()

if __name__ == "__main__":
    main()

File: eisenhower.cpp
// eisenhower.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <random>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

struct Task {
    string id;
    string title;
    string description;
    int quadrant;
    string created;
    bool completed;
};

class EisenhowerApp {
private:
    vector<Task> tasks;
    string dataFile = "tasks.json";

    string generateId() {
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> dis(0, 15);
        stringstream ss;
        for (int i = 0; i < 8; i++) {
            ss << hex << dis(gen);
        }
        return ss.str();
    }

    void load() {
        ifstream f(dataFile);
        if (!f.is_open()) return;
        json j;
        f >> j;
        for (auto& item : j) {
            Task t;
            t.id = item["id"];
            t.title = item["title"];
            t.description = item.value("description", "");
            t.quadrant = item["quadrant"];
            t.created = item.value("created", "");
            t.completed = item.value("completed", false);
            tasks.push_back(t);
        }
    }

    void save() {
        json j = json::array();
        for (auto& t : tasks) {
            json item;
            item["id"] = t.id;
            item["title"] = t.title;
            item["description"] = t.description;
            item["quadrant"] = t.quadrant;
            item["created"] = t.created;
            item["completed"] = t.completed;
            j.push_back(item);
        }
        ofstream f(dataFile);
        f << setw(2) << j << endl;
    }

public:
    EisenhowerApp() { load(); }

    void addTask(string title, string desc, int quadrant) {
        Task t;
        t.id = generateId();
        t.title = title;
        t.description = desc;
        t.quadrant = quadrant;
        time_t now = time(nullptr);
        t.created = string(ctime(&now));
        t.created.pop_back();
        t.completed = false;
        tasks.push_back(t);
        save();
        cout << "Task added: " << t.title << " (ID: " << t.id << ")" << endl;
    }

    void deleteTask(string id) {
        for (auto it = tasks.begin(); it != tasks.end(); ++it) {
            if (it->id == id) {
                tasks.erase(it);
                save();
                cout << "Task " << id << " deleted." << endl;
                return;
            }
        }
        cout << "Task " << id << " not found." << endl;
    }

    void moveTask(string id, int newQuadrant) {
        for (auto& t : tasks) {
            if (t.id == id) {
                t.quadrant = newQuadrant;
                save();
                cout << "Task " << id << " moved to quadrant " << newQuadrant << "." << endl;
                return;
            }
        }
        cout << "Task " << id << " not found." << endl;
    }

    void listTasks(int quadrant = 0) {
        vector<Task> filtered;
        for (auto& t : tasks) {
            if (quadrant == 0 || t.quadrant == quadrant) {
                filtered.push_back(t);
            }
        }
        if (filtered.empty()) {
            cout << "No tasks found." << endl;
            return;
        }
        string labels[] = {"Urgent & Important", "Not Urgent but Important",
                           "Urgent but Not Important", "Not Urgent & Not Important"};
        string colors[] = {"\033[31m", "\033[32m", "\033[33m", "\033[37m"};
        cout << "\nEisenhower Matrix" << endl;
        cout << string(60, '=') << endl;
        for (int q = 1; q <= 4; q++) {
            cout << colors[q-1] << labels[q-1] << "\033[0m" << endl;
            bool has = false;
            for (auto& t : filtered) {
                if (t.quadrant == q) {
                    cout << "  [" << t.id << "] " << t.title << " - "
                         << (t.description.empty() ? "No description" : t.description) << endl;
                    has = true;
                }
            }
            if (!has) cout << "  (empty)" << endl;
            if (q < 4) cout << string(40, '-') << endl;
        }
        cout << string(60, '=') << endl;
    }

    void stats() {
        int counts[5] = {0};
        for (auto& t : tasks) counts[t.quadrant]++;
        string labels[] = {"Urgent & Important", "Not Urgent but Important",
                           "Urgent but Not Important", "Not Urgent & Not Important"};
        cout << "Eisenhower Statistics" << endl;
        for (int q = 1; q <= 4; q++) {
            cout << labels[q-1] << ": " << counts[q] << " tasks" << endl;
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: eisenhower <command> [options]" << endl;
        return 1;
    }
    EisenhowerApp app;
    string cmd = argv[1];
    if (cmd == "add") {
        string title, desc = "";
        int quadrant = 1;
        for (int i = 2; i < argc; i++) {
            string arg = argv[i];
            if (arg == "-t" && i+1 < argc) title = argv[++i];
            else if (arg == "-d" && i+1 < argc) desc = argv[++i];
            else if (arg == "-q" && i+1 < argc) quadrant = stoi(argv[++i]);
        }
        app.addTask(title, desc, quadrant);
    } else if (cmd == "list") {
        int quadrant = 0;
        for (int i = 2; i < argc; i++) {
            if (string(argv[i]) == "-q" && i+1 < argc) quadrant = stoi(argv[++i]);
        }
        app.listTasks(quadrant);
    } else if (cmd == "move") {
        string id = "";
        int q = 0;
        for (int i = 2; i < argc; i++) {
            if (string(argv[i]) == "--id" && i+1 < argc) id = argv[++i];
            if (string(argv[i]) == "-q" && i+1 < argc) q = stoi(argv[++i]);
        }
        app.moveTask(id, q);
    } else if (cmd == "delete") {
        string id = "";
        for (int i = 2; i < argc; i++) {
            if (string(argv[i]) == "--id" && i+1 < argc) id = argv[++i];
        }
        app.deleteTask(id);
    } else if (cmd == "stats") {
        app.stats();
    } else {
        cerr << "Unknown command." << endl;
    }
    return 0;
}
